feat(tabs): make component responsive

jcfranco opened the pull request:
**Related Issue:** #6689 

## Summary

This improves how `tabs` works at narrow widths where `tab-title`s can't be displayed entirely. 

### Notable changes

* overflowing `tab-title`s will be clipped
* affordances to scroll clipped `tab-titles` forward/backward will be displayed accordingly
* selecting a `tab-title` that is partially clipped will be scrolled into view
* `tab-title`s can be scrolled via mousewheel

See [design spec](https://www.figma.com/file/JAINbGHSykI8JQCVzOHiSz/Tabs---responsive-design-%5B6689%5D?node-id=862%3A13837&mode=dev) for more details.

jcfranco commented on packages/calcite-components/.storybook/utils.tsx:
```diff
@@ -146,7 +146,7 @@ export const filterComponentAttributes = (
  * This helper creates a story that captures all breakpoints across all scales for testing.
  *
- * @param singleStoryHtml – HTML story template with placeholders for `scale` attributes (e.g., `{scale}`).
+ * @param singleStoryHtml – HTML story template with placeholders for `scale` attributes (e.g., `{scale}`). You can additionally use `.breakpoint-stories-container` and `.breakpoint-story-container` to style breakpoint story containers.
```
Extracted this over to https://github.com/Esri/calcite-design-system/pull/8739.

driskull commented on packages/calcite-components/src/components/tab-title/tab-title.scss:
```diff
@@ -26,29 +26,29 @@
-    padding-inline-start: 36px; //28px button width + 0.5rem padding
+  .scale-l {
+    .content {
+      padding-inline-start: 40px; //36px button width + .25 padding
```
Maybe to get rid of these comments we should have an internal var for button width and padding? Same goes for other uses below.

driskull commented on packages/calcite-components/src/components/tab-nav/tab-nav.tsx:
```diff
@@ -301,35 +412,197 @@ export class TabNav {
+      return;
+    }
+
+    this.intersectionObserver = createObserver("intersection", () => this.updateScrollingState(), {
```
Any chance `storeTabTitleWrapperRef` could be called again and this intersectionObserver may need to be disconnected before the new one is created?

driskull commented on packages/calcite-components/src/components/tab-nav/tab-nav.tsx:
```diff
@@ -301,35 +412,197 @@ export class TabNav {
+    readTask(() => {
+      const tabTitleContainer = this.tabTitleContainerEl;
+      const containerBounds = tabTitleContainer.getBoundingClientRect();
+      const tabTitles = Array.from(this.el.querySelectorAll("calcite-tab-title"));
```
maybe these could be stored with the onSlotChange and not queried via DOM?

driskull commented on packages/calcite-components/src/components/tab-nav/tab-nav.tsx:
```diff
@@ -414,9 +673,38 @@ export class TabNav {
+          [CSS.scrollBackwardContainerButton]: !isEnd,
+          [CSS.scrollForwardContainerButton]: isEnd,
+        }}
+        hidden={(isEnd && !hasOverflowingEndTabTitle) || (!isEnd && !hasOverflowingStartTabTitle)}
```
Instead of hiding, should it just not be rendered? Not a deal breaker just wondering.

driskull reviewed (approved):
Looks good! 🏆 ✨

macandcheese commented:
Nice! Some design / UX feedback mostly (cc @ashetland @SkyeSeitz)

- If I use the buttons via keyboard, then "shift tab back" to the tab items, and the now focused item is out of view, it does not scroll into view, so it is impossible to tell the tab has been successfully focused.

- Using the arrow keys, it seems the trigger to "scroll to focused item" is either a step off or happens to be below the width threshold sometimes. What this means is I've used the arrow key, navigated to a now hidden tab, and it hasn't re-adjusted yet. Navigating to another will scroll this one into view, so maybe there needs to be a width buffer etc.

- If I have navigated with the arrow keys away from the active tab to focus-but-not-active a sibling, and then use Tab, the focus doesn't go to the arrow (or content). Not sure if intentional but it does re-focus the active tab before allowing you to tab again through.

- Using the arrow buttons (or arrow keys) doesn't always increment a full title into view. Sometimes I can hit "next" or "prev" 2/3 times and a long title will still be in view - can we ensure a new tab is visible?

- Can we use a rule to make scrolling respond better? Right now I need to swipe-swipe-swipe and it doesn't seem to have inertia scrolling making it tough to use on a phone or trackpad / mouse.

- Should we support long-press on these arrow keys in any way on mobile?

Style issues:

- This focus area is tabbable - if this is intentional we should use a default focus style:
- Is it a bug that there is no border on top of these arrow buttons?
![Screenshot 2024-02-13 at 2 08 09 PM](https://github.com/Esri/calcite-design-system/assets/4733155/c9405f3d-6437-469d-9095-a146a1d0a939)

- Design - do we want to limit these to squares? These are taking up too much space IMO:
- Design - is this the correct spacing here? It seems really cramped and spacious at the same time.
![Screenshot 2024-02-13 at 2 07 24 PM](https://github.com/Esri/calcite-design-system/assets/4733155/f9409eb9-7763-4090-b19a-3d6180e5d58a)

- The focus area outline is misaligned:
![Screenshot 2024-02-13 at 2 07 03 PM](https://github.com/Esri/calcite-design-system/assets/4733155/e49b2650-04fc-46b1-9613-86399c94446b)

- The padding on some of these is a bit weird - compare the first tabs to the subsequent ones.
![Screenshot 2024-02-13 at 2 06 29 PM](https://github.com/Esri/calcite-design-system/assets/4733155/540482c8-550a-4c7e-a6b0-03f90684afba)

- I suggest we get rid of the "zoom border animation", it is adding a bit of confusion to a crowded area (IMO). Just render the active state visual with a border or inset box-shadow.

driskull commented:
It might be nice if instead of having to horizontally scroll through tabs there was an overflow menu. Not a huge fan of having to horizontally "carousel" through tabs.

![image](https://github.com/Esri/calcite-design-system/assets/1231455/b1dc102b-3802-4e74-9b0c-563b1e8ebc62)

SkyeSeitz commented:
> It might be nice if instead of having to horizontally scroll through tabs there was an overflow menu. Not a huge fan of having to horizontally "carousel" through tabs.
> 
> ![image](https://private-user-images.githubusercontent.com/1231455/304571157-b1dc102b-3802-4e74-9b0c-563b1e8ebc62.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDc4NjUzNjYsIm5iZiI6MTcwNzg2NTA2NiwicGF0aCI6Ii8xMjMxNDU1LzMwNDU3MTE1Ny1iMWRjMTAyYi0zODAyLTRlNzQtOWIwYy01NjNiMWU4ZWJjNjIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDIxMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDAyMTNUMjI1NzQ2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDUxNDBmNzk4MjM2NjZkMDdlMGY0N2MyNWQ2NDI4MjZlOTZiNTk0MzhkMzliNWE2YWUyYWY1OTk5N2I0NjdhZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.hM8jGCt6GQqoZontqLUkanVi73EzHjTvRfGYd97_Tg0)

Might be subjective, but I think an overflow menu is less expected for Tabs specifically. See below for an example with the scroll function working..

https://github.com/Esri/calcite-design-system/assets/29716057/cdb66601-d8f0-4359-b0da-4b98cceea986

SkyeSeitz commented:
Agree with all Adam's points above and have similar thoughts (tried to not repeat any notes, but apologies if there is overlap)
* > This focus area is tabbable - if this is intentional we should use a default focus style:

Referring to the content area, correct? I had the same question, although unless there is something clickable in the content area, I am not sure it needs to be focusable. Looks like this is how it works & is styled currently, however. 
    
* > do we want to limit these to squares? These are taking up too much space IMO:

Agreed. These should have the following square dimensions
<img width="1665" alt="image" src="https://github.com/Esri/calcite-design-system/assets/29716057/efc8f389-0fb2-4c50-88a0-f28cda0b79b0">


Design UX questions: 
* When arrowing through tabs, I am not sure the last tab needs to roll into the 1st tab? Thoughts?
* When keyboard navigating to one of the ends of Tab via the chevron buttons, the focus ring goes away after reaching the end. I think I would expect it to move to the selected tab when the focused button goes away?

jcfranco commented:
> * Using the arrow buttons (or arrow keys) doesn't always increment a full title into view. 
Sometimes I can hit "next" or "prev" 2/3 times and a long title will still be in view - can we ensure a new tab is visible?
> * If I use the buttons via keyboard, then "shift tab back" to the tab items, and the now focused item is out of view, it does not scroll into view, so it is impossible to tell the tab has been successfully focused.
> * Using the arrow keys, it seems the trigger to "scroll to focused item" is either a step off or happens to be below the width threshold sometimes. What this means is I've used the arrow key, navigated to a now hidden tab, and it hasn't re-adjusted yet. Navigating to another will scroll this one into view, so maybe there needs to be a width buffer etc.
> * Is it a bug that there is no border on top of these arrow buttons?
> * Design - do we want to limit these to squares? These are taking up too much space IMO:

Fixed! ✨🔨✨

> * Can we use a rule to make scrolling respond better? Right now I need to swipe-swipe-swipe and it doesn't seem to have inertia scrolling making it tough to use on a phone or trackpad / mouse.

Updated swipe-scrolling to be ↕️on desktop and ↔️on mobile.

> * Should we support long-press on these arrow keys in any way on mobile?

The updated swipe-scrolling might take care of this now.

> * If I have navigated with the arrow keys away from the active tab to focus-but-not-active a sibling, and then use Tab, the focus doesn't go to the arrow (or content). Not sure if intentional but it does re-focus the active tab before allowing you to tab again through.

We could refine this if needed. The roving tab helper might help here.

> * This focus area is tabbable - if this is intentional we should use a default focus style:
> Referring to the content area, correct? I had the same question, although unless there is something clickable in the content area, I am not sure it needs to be focusable. Looks like this is how it works & is styled currently, however.

This is by intentional for a11y purposes, there is an issue to revisit making it tabbable only when interactive (https://github.com/Esri/calcite-design-system/issues/7576).

> * The focus area outline is misaligned:

Is this specific to the close button? If so, we can fix this separately since this issue is present in production too (https://github.com/Esri/calcite-design-system/issues/8772).

> * The padding on some of these is a bit weird - compare the first tabs to the subsequent ones.

Updated spacing to better match the spec. There is an issue though with center + closable layouts when they're used in narrow widths. You can see this in the UI kit:

![Screenshot 2024-02-16 at 4 52 21 PM](https://github.com/Esri/calcite-design-system/assets/197440/f3c9daee-7ac7-4116-8008-582f2a6ea6c7)

cc @SkyeSeitz @ashetland

> * I suggest we get rid of the "zoom border animation", it is adding a bit of confusion to a crowded area (IMO). Just render the active state visual with a border or inset box-shadow.

I'm down to remove/simplify this. Do you think we could tackle it as a separate issue or is it critical for this PR?

> * When arrowing through tabs, I am not sure the last tab needs to roll into the 1st tab? Thoughts?

Do you mean using keyboard left/right arrows when focus is on a tab or clicking on the prev/next tab actions? If it's the latter, this behavior is specified under Q.3 under `"Franco questions 01.17.24"` from the responsive tabs spec.

> * When keyboard navigating to one of the ends of Tab via the chevron buttons, the focus ring goes away after reaching the end. I think I would expect it to move to the selected tab when the focused button goes away? 

This might be confusing since you'll scroll back to the selected tab after 'paging' all the way to the end. E.g., selected tab is 2nd one, you page via keyboard with the chevron button until the end, it scrolls back to the 2nd one. I might be missing something though.

macandcheese commented:
Works a lot better now!

>This is by intentional for a11y purposes, there is an issue to revisit making it tabbable only when interactive (https://github.com/Esri/calcite-design-system/issues/7576).
Maybe we can use a shared focus style so it isn't browser default?

>I'm down to remove/simplify this. Do you think we could tackle it as a separate issue or is it critical for this PR?
Separate is good, it might resolve some of the alignment items, but it would be a refactor to remove associated code.

>This might be confusing since you'll scroll back to the selected tab after 'paging' all the way to the end. E.g., selected tab is 2nd one, you page via keyboard with the chevron button until the end, it scrolls back to the 2nd one. I might be missing something though.

Maybe the arrow buttons do not need to be in the tab order. They are likely to be used via tap / click. But, to a keyboard-only user, using those doesn't provide any use - it's just moving what is visible, not selecting a new tab. Keyboard-only users can already do this more efficiently with arrow keys. 

So, I think we could exclude them from tab order. This would be similar to some input functionality like number spinner, clear button - that are clickable entities but excluded from tab focus stack because the keyboard-only user has better usability without them.

SkyeSeitz commented:
1. >So, I think we could exclude them from tab order. This would be similar to some input functionality like number spinner, clear button - that are clickable entities but excluded from tab focus stack because the keyboard-only user has better usability without them.

This is a super good point and would take care of my question. 

2. >Updated swipe-scrolling to be ↕️on desktop and ↔️on mobile.

Is there a reason ↔️ does not scroll on desktop? I would expect both directions to scroll.

jcfranco commented:
> Maybe we can use a shared focus style so it isn't browser default?

Agreed. Created https://github.com/Esri/calcite-design-system/issues/8799 for this.

> Separate is good, it might resolve some of the alignment items, but it would be a refactor to remove associated code.

I will create an issue for this today.

> Maybe the arrow buttons do not need to be in the tab order. 

Removed from tab order! I think this works way better for keyboard usage (IMO).

> Is there a reason ↔️ does not scroll on desktop? I would expect both directions to scroll.

When testing locally, it seems when you scroll with the mouse wheel and trackpad (at least mine), I get both delta x and delta y values. I wouldn't know which one to use if I handle both horizontal/vertical scrolling.

driskull commented on packages/calcite-components/src/components/tab-nav/tab-nav.scss:
```diff
@@ -1,78 +1,199 @@
+  }
+}
+
+.tab-nav::-webkit-scrollbar {
```
Whats this for? we're hiding the scrollbar but still allowing scrolling?

jcfranco replied:
It was added [here](https://github.com/Esri/calcite-design-system/commit/712fb31655e851973502737764588881b73b4e3c). Removing it doesn't seem to have any effect. @Elijbet Do you recall why this is needed? If not, I can drop this. Good catch! ✨🎣✨

driskull reviewed (approved):
👍

jcfranco commented:
> Is there a reason ↔️ does not scroll on desktop? I would expect both directions to scroll.

Managed to make this work thanks to https://github.com/react-component/tabs. 🎉

jcfranco commented:
This is now officially good to land, but the extent of changes aren't low-risk. I suggest we install this as soon tomorrow's maintenance release is wrapped up. cc @geospatialem @brittneytewks
